packages/apollo-link/src/link.ts

```diff
@@ -8,179 +10,123 @@ import {
+  left: ApolloLink | RequestHandler,
+  right: ApolloLink | RequestHandler = new ApolloLink(passthrough),
+): ApolloLink => {
+  const leftLink = toLink(left);
+  const rightLink = toLink(right);
+
+  if (isTerminating(leftLink) && isTerminating(rightLink)) {
+    return new ApolloLink(operation => {
+      return test(operation)
+        ? leftLink.request(operation) || Observable.of()
+        : rightLink.request(operation) || Observable.of();
+    });
+  } else {
+    return new ApolloLink((operation, forward) => {
+      return test(operation)
+        ? leftLink.request(operation, forward) || Observable.of()
+        : rightLink.request(operation, forward) || Observable.of();
+    });
+  }
+};
+
+// join two Links together
+export const concat = (
+  first: ApolloLink | RequestHandler,
+  second: ApolloLink | RequestHandler,
+) => {
+  const firstLink = toLink(first);
+  if (isTerminating(firstLink)) {
+    console.warn(
+      new LinkError(
+        `You are calling concat on a terminating link, which will have no effect`,
+        firstLink,
+      ),
     );
+    return firstLink;
   }
+  const nextLink = toLink(second);
+
+  if (isTerminating(nextLink)) {
+    return new ApolloLink(
+      operation =>
+        firstLink.request(
+          operation,
+          op => nextLink.request(op) || Observable.of(),
+        ) || Observable.of(),
+    );
+  } else {
+    return new ApolloLink((operation, forward) => {
+      return (
+        firstLink.request(operation, op => {
+          return nextLink.request(op, forward) || Observable.of();
+        }) || Observable.of()
+      );
+    });
+  }
+};
 
-  // split allows for creating a split point in an execution chain
-  // like filter, it can be used to direct operations based
-  // on request information. Instead of dead ending an execution,
-  // split allows for new chains to be formed.
-  public static split(
-    test: (op: Operation) => boolean,
-    left: ApolloLink | RequestHandler,
-    right: ApolloLink | RequestHandler = ApolloLink.passthrough(),
-  ): ApolloLink {
-    const leftLink = toLink(left);
-    const rightLink = toLink(right);
-
-    if (isTerminating(leftLink) && isTerminating(rightLink)) {
-      return new FunctionLink(operation => {
-        return test(operation)
-          ? leftLink.request(operation) || Observable.of()
-          : rightLink.request(operation) || Observable.of();
-      });
-    } else {
-      return new FunctionLink((operation, forward) => {
-        return test(operation)
-          ? leftLink.request(operation, forward) || Observable.of()
-          : rightLink.request(operation, forward) || Observable.of();
-      });
-    }
+// XXX add in type support for context manipulation
+// XXX remove mutability of context
+export class ApolloLink {
+  constructor(request?: RequestHandler) {
+    if (request) this.request = request;
   }
 
+  public static empty = empty;
+  // XXX should we deprecate from in favor of fold?
```

Comment:
`fold` makes me feel like I'm doing functional programming - honestly it makes me uncomfortable.

Reply:
Could it be `concat`?

Reply:
I'll revert it back to from. I initially thought this would be a problem in case `from` became part of the lexical grammar (with `import Foo from 'bar'`), but for now according to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar its not so I think its okay to use.

Agreed on fold